File: server/src/controllers/movementsController.js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

const movementInclude = {
    item: { select: { id: true, sku: true, name: true } },
    warehouse: { select: { id: true, name: true } },
    fromWarehouse: { select: { id: true, name: true } },
    toWarehouse: { select: { id: true, name: true } },
    user: { select: { id: true, name: true, email: true, role: true } },
};

const getMovements = async (req, res, next) => {
    try {
        const { warehouse, item, type, from, to, page = 1, limit = 20 } = req.query;
        const where = {};
        if (warehouse) where.warehouseId = parseInt(warehouse);
        if (item) where.itemId = parseInt(item);
        if (type) where.movementType = type;
        if (from || to) {
            where.timestamp = {};
            if (from) where.timestamp.gte = new Date(from);
            if (to) where.timestamp.lte = new Date(to + 'T23:59:59.999Z');
        }

        const skip = (parseInt(page) - 1) * parseInt(limit);
        const [total, movements] = await Promise.all([
            prisma.stockMovement.count({ where }),
            prisma.stockMovement.findMany({
                where,
                include: movementInclude,
                orderBy: { timestamp: 'desc' },
                skip,
                take: parseInt(limit),
            }),
        ]);

        res.json({ total, page: parseInt(page), limit: parseInt(limit), data: movements });
    } catch (err) {
        next(err);
    }
};

const getAudit = async (req, res, next) => {
    try {
        const { from, to, userId, warehouseId, page = 1, limit = 20 } = req.query;
        const where = {};
        if (userId) where.userId = parseInt(userId);
        if (warehouseId) where.warehouseId = parseInt(warehouseId);
        if (from || to) {
            where.timestamp = {};
            if (from) where.timestamp.gte = new Date(from);
            if (to) where.timestamp.lte = new Date(to + 'T23:59:59.999Z');
        }

        const skip = (parseInt(page) - 1) * parseInt(limit);
        const [total, logs] = await Promise.all([
            prisma.stockMovement.count({ where }),
            prisma.stockMovement.findMany({
                where,
                include: movementInclude,
                orderBy: { timestamp: 'desc' },
                skip,
                take: parseInt(limit),
            }),
        ]);

        res.json({ total, page: parseInt(page), limit: parseInt(limit), data: logs });
    } catch (err) {
        next(err);
    }
};

const getDashboardStats = async (req, res, next) => {
    try {
        const [totalItems, totalStockResult, lowStockCount, pendingTransfers, recentMovements, warehouseStocks, last7DaysMovements] = await Promise.all([
            prisma.inventoryItem.count(),
            prisma.stockLevel.aggregate({ _sum: { quantity: true } }),
            prisma.stockLevel.count({ where: { quantity: { lte: 100 } } }),
            prisma.stockMovement.count({ where: { movementType: { in: ['TRANSFER_IN', 'TRANSFER_OUT'] } } }),
            prisma.stockMovement.findMany({
                take: 10,
                orderBy: { timestamp: 'desc' },
                include: movementInclude,
            }),
            prisma.warehouse.findMany({
                include: { stockLevels: true },
            }),
            prisma.stockMovement.findMany({
                where: { timestamp: { gte: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000) } },
                orderBy: { timestamp: 'asc' },
                select: { timestamp: true, movementType: true, quantity: true },
            }),
        ]);

        const warehouseStockData = warehouseStocks.map((wh) => ({
            name: wh.name,
            totalStock: wh.stockLevels.reduce((sum, sl) => sum + sl.quantity, 0),
        }));

        // build daily movement totals for chart
        const dailyMap = {};
        last7DaysMovements.forEach((mv) => {
            const day = mv.timestamp.toISOString().split('T')[0];
            if (!dailyMap[day]) dailyMap[day] = { date: day, added: 0, removed: 0, transferred: 0 };
            if (mv.movementType === 'ADD') dailyMap[day].added += mv.quantity;
            else if (mv.movementType === 'REMOVE') dailyMap[day].removed += mv.quantity;
            else dailyMap[day].transferred += mv.quantity;
        });
        const dailyMovements = Object.values(dailyMap).sort((a, b) => a.date.localeCompare(b.date));

        res.json({
            kpi: {
                totalItems,
                totalStock: totalStockResult._sum.quantity || 0,
                lowStockAlerts: lowStockCount,
                pendingTransfers,
            },
            warehouseStockData,
            dailyMovements,
            recentMovements,
        });
    } catch (err) {
        next(err);
    }
};

module.exports = { getMovements, getAudit, getDashboardStats };

File: server_java/src/main/java/com/terragrid/controller/MovementsController.java
package com.terragrid.controller;

import com.terragrid.dto.DashboardResponse;
import com.terragrid.model.MovementType;
import com.terragrid.model.StockMovement;
import com.terragrid.model.Warehouse;
import com.terragrid.repository.InventoryItemRepository;
import com.terragrid.repository.StockLevelRepository;
import com.terragrid.repository.StockMovementRepository;
import com.terragrid.repository.WarehouseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/movements")
public class MovementsController {

    @Autowired
    private StockMovementRepository movementRepository;

    @Autowired
    private InventoryItemRepository itemRepository;

    @Autowired
    private StockLevelRepository stockLevelRepository;

    @Autowired
    private WarehouseRepository warehouseRepository;

    @GetMapping
    public ResponseEntity<Map<String, Object>> getMovements(
            @RequestParam(required = false) Integer warehouse,
            @RequestParam(required = false) Integer item,
            @RequestParam(required = false) MovementType type,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        Pageable pageable = PageRequest.of(page - 1, limit, Sort.by("timestamp").descending());
        Page<StockMovement> result = movementRepository.findFilteredMovements(null, warehouse, item, type, from, to, pageable);

        Map<String, Object> response = new HashMap<>();
        response.put("total", result.getTotalElements());
        response.put("page", page);
        response.put("limit", limit);
        response.put("data", result.getContent());

        return ResponseEntity.ok(response);
    }

    @GetMapping("/dashboard")
    public ResponseEntity<DashboardResponse> getDashboardStats() {
        Long totalItems = itemRepository.count();
        Long totalStock = stockLevelRepository.sumTotalQuantity();
        Long lowStockAlerts = stockLevelRepository.countLowStock();
        Long pendingTransfers = movementRepository.countByMovementTypeIn(Arrays.asList(MovementType.TRANSFER_IN, MovementType.TRANSFER_OUT));

        List<Warehouse> warehouses = warehouseRepository.findAll();
        List<DashboardResponse.WarehouseStockData> warehouseStockData = warehouses.stream().map(wh -> {
            Long whStock = stockLevelRepository.findByWarehouseId(wh.getId()).stream()
                    .mapToLong(sl -> (long) sl.getQuantity())
                    .sum();
            return DashboardResponse.WarehouseStockData.builder()
                    .name(wh.getName())
                    .totalStock(whStock)
                    .build();
        }).collect(Collectors.toList());

        LocalDateTime sevenDaysAgo = LocalDateTime.now().minusDays(7);
        List<StockMovement> movements = movementRepository.findByTimestampAfter(sevenDaysAgo);
        
        Map<String, DashboardResponse.DailyMovementData> dailyMap = new TreeMap<>();
        movements.forEach(m -> {
            String date = m.getTimestamp().toLocalDate().toString();
            DashboardResponse.DailyMovementData data = dailyMap.getOrDefault(date, 
                    DashboardResponse.DailyMovementData.builder().date(date).added(0).removed(0).transferred(0).build());
            
            if (m.getMovementType() == MovementType.ADD) data.setAdded(data.getAdded() + m.getQuantity());
            else if (m.getMovementType() == MovementType.REMOVE) data.setRemoved(data.getRemoved() + m.getQuantity());
            else data.setTransferred(data.getTransferred() + m.getQuantity());
            
            dailyMap.put(date, data);
        });

        return ResponseEntity.ok(DashboardResponse.builder()
                .kpi(DashboardResponse.KpiData.builder()
                        .totalItems(totalItems)
                        .totalStock(totalStock != null ? totalStock : 0L)
                        .lowStockAlerts(lowStockAlerts)
                        .pendingTransfers(pendingTransfers)
                        .build())
                .warehouseStockData(warehouseStockData)
                .dailyMovements(new ArrayList<>(dailyMap.values()))
                .recentMovements(movementRepository.findTop10ByOrderByTimestampDesc())
                .build());
    }

    @GetMapping("/audit")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Map<String, Object>> getAudit(
            @RequestParam(required = false) Integer userId,
            @RequestParam(required = false) Integer warehouseId,
            @RequestParam(required = false) Integer itemId,
            @RequestParam(required = false) MovementType type,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        Pageable pageable = PageRequest.of(page - 1, limit, Sort.by("timestamp").descending());
        Page<StockMovement> result = movementRepository.findFilteredMovements(userId, warehouseId, itemId, type, from, to, pageable);

        Map<String, Object> response = new HashMap<>();
        response.put("total", result.getTotalElements());
        response.put("page", page);
        response.put("limit", limit);
        response.put("data", result.getContent());

        return ResponseEntity.ok(response);
    }
}
